## Importing Libraries

In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

### Loading Dataset

In [14]:
data = pd.read_csv("E:/Russian_Housing_Dataset1.csv")
print(data.shape)
data.head()


(99999, 272)


full_sq  life_sq  floor product_type           sub_area       area_m  \
0     43.0     27.0    4.0   Investment           Bibirevo  6407578.100   
1     34.0     19.0    3.0   Investment  Nagatinskij Zaton  9589336.912   
2     43.0     29.0    2.0   Investment     Tekstil'shhiki  4808269.831   
3     77.0     77.0    4.0   Investment          Basmannoe  8398460.622   
4     67.0     46.0   14.0   Investment     Nizhegorodskoe  7506452.020   

   raion_popul  green_zone_part  indust_part  children_preschool  ...  \
0     155572.0         0.189727     0.000070              9576.0  ...   
1     115352.0         0.372602     0.049637              6880.0  ...   
2     101708.0         0.112560     0.118537              5879.0  ...   
3     108171.0         0.015234     0.037316              5706.0  ...   
4      43795.0         0.007670     0.486246              2418.0  ...   

   cafe_count_5000_price_2500  cafe_count_5000_price_4000  \
0                         9.0                         4.0   
1                        15.0                         3.0   
2                        10.0                         3.0   
3                       319.0                       108.0   
4                        62.0                        14.0   

   cafe_count_5000_price_high  big_church_count_5000  church_count_5000  \
0                         0.0                   13.0               22.0   
1                         0.0                   15.0               29.0   
2                         0.0                   11.0               27.0   
3                        17.0                  135.0              236.0   
4                         1.0                   53.0               78.0   

   mosque_count_5000  leisure_count_5000  sport_count_5000 market_count_5000  \
0                1.0                 0.0              52.0               4.0   
1                1.0                10.0              66.0              14.0   
2                0.0                 4.0              67.0              10.0   
3                2.0                91.0             195.0              14.0   
4                1.0                20.0             113.0              17.0   

    price_doc  
0   5850000.0  
1   6000000.0  
2   5700000.0  
3  16331452.0  
4   9100000.0  

[5 rows x 272 columns]

## Data Pre Processing

In [15]:
data.describe()

full_sq       life_sq         floor        area_m    raion_popul  \
count  99999.000000  99999.000000  99999.000000  9.999900e+04   99999.000000   
mean     450.974601    596.342883     11.916919  2.570084e+07  103666.321411   
std     1115.307106   1535.024315     14.761103  4.171087e+07   51981.445433   
min        0.000000      0.000000      0.000000  2.081628e+06    2546.000000   
25%       40.000000     23.082199      4.000000  7.128794e+06   76284.000000   
50%       52.516018     32.529567      7.000000  9.629358e+06  101982.000000   
75%       75.988433     53.000000     12.000000  1.675112e+07  130314.905350   
max     5326.000000   7478.000000     76.998630  2.060718e+08  247469.000000   

       green_zone_part   indust_part  children_preschool  \
count     99999.000000  99999.000000        99999.000000   
mean          0.238107      0.145010         6436.288113   
std           0.200461      0.138731         3816.536558   
min           0.001879      0.000000          175.000000   
25%           0.067927      0.035145         4215.000000   
50%           0.183969      0.101872         5797.000000   
75%           0.372602      0.238617         7751.000000   
max           0.852923      0.521867        19223.000000   

       preschool_education_centers_raion  children_school  ...  \
count                       99999.000000     99999.000000  ...   
mean                            5.059566      6650.794929  ...   
std                             2.740557      3876.750082  ...   
min                             0.000000       168.000000  ...   
25%                             4.000000      4439.000000  ...   
50%                             5.000000      5992.000000  ...   
75%                             6.330909      7976.000000  ...   
max                            13.000000     19083.000000  ...   

       cafe_count_5000_price_2500  cafe_count_5000_price_4000  \
count                99999.000000                99999.000000   
mean                    59.600206                   21.534283   
std                     93.928443                   37.211003   
min                      0.000000                    0.000000   
25%                      7.000000                    1.386062   
50%                     13.979400                    3.000000   
75%                     61.485717                   17.000000   
max                    377.000000                  147.000000   

       cafe_count_5000_price_high  big_church_count_5000  church_count_5000  \
count                99999.000000           99999.000000       99999.000000   
mean                     3.860331              26.900600          48.747126   
std                      7.478604              37.518277          60.046671   
min                      0.000000               0.000000           0.000000   
25%                      0.000000               6.000000          13.082211   
50%                      0.000000              10.000000          24.000000   
75%                      2.801950              23.000000          45.624406   
max                     30.000000             151.000000         250.000000   

       mosque_count_5000  leisure_count_5000  sport_count_5000  \
count       99999.000000        99999.000000      99999.000000   
mean            0.547042           16.655833         70.699255   
std             0.607492           26.389935         48.485232   
min             0.000000            0.000000          0.000000   
25%             0.000000            1.000000         39.056711   
50%             0.230253            4.089280         61.443864   
75%             1.000000           13.805097         93.413009   
max             2.000000          106.000000        218.000000   

       market_count_5000     price_doc  
count       99999.000000  9.999900e+04  
mean            7.933014  1.498095e+07  
std             5.048946  2.181344e+07  
min             0.000000  1.000000e+05  
25%             4.000000  5.469147e+06  
50%             

In [16]:
data.dtypes

full_sq               float64
life_sq               float64
floor                 float64
product_type           object
sub_area               object
                       ...   
mosque_count_5000     float64
leisure_count_5000    float64
sport_count_5000      float64
market_count_5000     float64
price_doc             float64
Length: 272, dtype: object

Listing all the Columns of the data

In [17]:
cols = data.columns
print(cols)

Index(['full_sq', 'life_sq', 'floor', 'product_type', 'sub_area', 'area_m',
       'raion_popul', 'green_zone_part', 'indust_part', 'children_preschool',
       ...
       'cafe_count_5000_price_2500', 'cafe_count_5000_price_4000',
       'cafe_count_5000_price_high', 'big_church_count_5000',
       'church_count_5000', 'mosque_count_5000', 'leisure_count_5000',
       'sport_count_5000', 'market_count_5000', 'price_doc'],
      dtype='object', length=272)


Listing Only Numerical Columns

In [18]:
num_cols = data._get_numeric_data().columns
print(num_cols)

Index(['full_sq', 'life_sq', 'floor', 'area_m', 'raion_popul',
       'green_zone_part', 'indust_part', 'children_preschool',
       'preschool_education_centers_raion', 'children_school',
       ...
       'cafe_count_5000_price_2500', 'cafe_count_5000_price_4000',
       'cafe_count_5000_price_high', 'big_church_count_5000',
       'church_count_5000', 'mosque_count_5000', 'leisure_count_5000',
       'sport_count_5000', 'market_count_5000', 'price_doc'],
      dtype='object', length=257)


Listing Only Categorical Columns

In [19]:

categorical_colums = list(set(cols) - set(num_cols))
categorical_colums

['culture_objects_top_25',
 'incineration_raion',
 'product_type',
 'thermal_power_plant_raion',
 'radiation_raion',
 'railroad_1line',
 'railroad_terminal_raion',
 'detention_facility_raion',
 'oil_chemistry_raion',
 'ecology',
 'big_market_raion',
 'sub_area',
 'water_1line',
 'nuclear_reactor_raion',
 'big_road1_1line']

In [20]:
data.shape

(99999, 272)

The Function below will convert all the string field with binary class "Yes" and "No" to 1 and 0.

In [21]:
def pre_process(data):
    data['radiation_raion'] = data['radiation_raion'].map({'yes': 1, 'no': 0})
    data['radiation_raion'] = pd.to_numeric(data['radiation_raion'])
    data['big_market_raion'] = data['big_market_raion'].map({'yes': 1, 'no': 0})
    data['big_market_raion'] = pd.to_numeric(data['big_market_raion'])
    data['ecology'] = data['ecology'].map({'satisfactory': 1, 'poor': 0, 'excellent': 3, 'good':2})
    data['ecology'] = pd.to_numeric(data['ecology'])
    data['oil_chemistry_raion'] = data['oil_chemistry_raion'].map({'yes': 1, 'no': 0})
    data['oil_chemistry_raion'] = pd.to_numeric(data['oil_chemistry_raion'])
    data['culture_objects_top_25'] = data['culture_objects_top_25'].map({'yes': 1, 'no': 0})
    data['culture_objects_top_25'] = pd.to_numeric(data['culture_objects_top_25'])
    data['detention_facility_raion'] = data['detention_facility_raion'].map({'yes': 1, 'no': 0})
    data['detention_facility_raion'] = pd.to_numeric(data['detention_facility_raion'])
    data['railroad_1line'] = data['railroad_1line'].map({'yes': 1, 'no': 0})
    data['railroad_1line'] = pd.to_numeric(data['railroad_1line'])
    data['incineration_raion'] = data['incineration_raion'].map({'yes': 1, 'no': 0})
    data['incineration_raion'] = pd.to_numeric(data['incineration_raion'])
    data['nuclear_reactor_raion'] = data['nuclear_reactor_raion'].map({'yes': 1, 'no': 0})
    data['nuclear_reactor_raion'] = pd.to_numeric(data['nuclear_reactor_raion'])
    data['thermal_power_plant_raion'] = data['thermal_power_plant_raion'].map({'yes': 1, 'no': 0})
    data['thermal_power_plant_raion'] = pd.to_numeric(data['thermal_power_plant_raion'])
    data['big_road1_1line'] = data['big_road1_1line'].map({'yes': 1, 'no': 0})
    data['big_road1_1line'] = pd.to_numeric(data['big_road1_1line'])
    data['big_road1_1line'] = data['big_road1_1line'].map({'yes': 1, 'no': 0})
    data['big_road1_1line'] = pd.to_numeric(data['big_road1_1line'])
    data['railroad_terminal_raion'] = data['railroad_terminal_raion'].map({'yes': 1, 'no': 0})
    data['railroad_terminal_raion'] = pd.to_numeric(data['railroad_terminal_raion'])
    data['product_type'] = data['product_type'].map({'Investment': 1, 'OwnerOccupier': 0})
    data['product_type'] = pd.to_numeric(data['product_type'])
    data['water_1line'] = data['water_1line'].map({'yes': 1, 'no': 0})
    data['water_1line'] = pd.to_numeric(data['water_1line'])

    #data.drop('row ID',axis=1,inplace=True)
    data.drop('sub_area',axis=1,inplace=True)

    return data

Calling Function

In [22]:
data = pre_process(data)
data.shape

(99999, 271)

Checking Null Values

In [23]:
#Checking Null Values.
data.isnull().sum()

full_sq               0
life_sq               0
floor                 0
product_type          0
area_m                0
                     ..
mosque_count_5000     0
leisure_count_5000    0
sport_count_5000      0
market_count_5000     0
price_doc             0
Length: 271, dtype: int64

Filling Null Values with Mean Value of column so that distrivution remains same.

In [24]:
data['price_doc'] = data['price_doc'].fillna((data['price_doc'].mean()))

In [25]:
#Checking Null Values again after dropping Null Values.
data.isnull().sum()

full_sq               0
life_sq               0
floor                 0
product_type          0
area_m                0
                     ..
mosque_count_5000     0
leisure_count_5000    0
sport_count_5000      0
market_count_5000     0
price_doc             0
Length: 271, dtype: int64

Filling Null values with 1 if there is any left.

In [26]:
data = data.fillna(1)

In [27]:
cols = data.columns
print(cols)

Index(['full_sq', 'life_sq', 'floor', 'product_type', 'area_m', 'raion_popul',
       'green_zone_part', 'indust_part', 'children_preschool',
       'preschool_education_centers_raion',
       ...
       'cafe_count_5000_price_2500', 'cafe_count_5000_price_4000',
       'cafe_count_5000_price_high', 'big_church_count_5000',
       'church_count_5000', 'mosque_count_5000', 'leisure_count_5000',
       'sport_count_5000', 'market_count_5000', 'price_doc'],
      dtype='object', length=271)


In [28]:
num_cols = data._get_numeric_data().columns
print(num_cols)

Index(['full_sq', 'life_sq', 'floor', 'product_type', 'area_m', 'raion_popul',
       'green_zone_part', 'indust_part', 'children_preschool',
       'preschool_education_centers_raion',
       ...
       'cafe_count_5000_price_2500', 'cafe_count_5000_price_4000',
       'cafe_count_5000_price_high', 'big_church_count_5000',
       'church_count_5000', 'mosque_count_5000', 'leisure_count_5000',
       'sport_count_5000', 'market_count_5000', 'price_doc'],
      dtype='object', length=271)


In [29]:

categorical_colums = list(set(cols) - set(num_cols))
categorical_colums

[]

As you can see above after pre processing all our Categorical columns are converted into Numerical.

In [30]:
data.head()

full_sq  life_sq  floor  product_type       area_m  raion_popul  \
0     43.0     27.0    4.0             1  6407578.100     155572.0   
1     34.0     19.0    3.0             1  9589336.912     115352.0   
2     43.0     29.0    2.0             1  4808269.831     101708.0   
3     77.0     77.0    4.0             1  8398460.622     108171.0   
4     67.0     46.0   14.0             1  7506452.020      43795.0   

   green_zone_part  indust_part  children_preschool  \
0         0.189727     0.000070              9576.0   
1         0.372602     0.049637              6880.0   
2         0.112560     0.118537              5879.0   
3         0.015234     0.037316              5706.0   
4         0.007670     0.486246              2418.0   

   preschool_education_centers_raion  ...  cafe_count_5000_price_2500  \
0                                5.0  ...                         9.0   
1                                5.0  ...                        15.0   
2                                4.0  ...                        10.0   
3                                7.0  ...                       319.0   
4                                2.0  ...                        62.0   

   cafe_count_5000_price_4000  cafe_count_5000_price_high  \
0                         4.0                         0.0   
1                         3.0                         0.0   
2                         3.0                         0.0   
3                       108.0                        17.0   
4                        14.0                         1.0   

   big_church_count_5000  church_count_5000  mosque_count_5000  \
0                   13.0               22.0                1.0   
1                   15.0               29.0                1.0   
2                   11.0               27.0                0.0   
3                  135.0              236.0                2.0   
4                   53.0               78.0                1.0   

   leisure_count_5000  sport_count_5000  market_count_5000   price_doc  
0                 0.0              52.0                4.0   5850000.0  
1                10.0              66.0               14.0   6000000.0  
2                 4.0              67.0               10.0   5700000.0  
3                91.0             195.0               14.0  16331452.0  
4                20.0             113.0               17.0   9100000.0  

[5 rows x 271 columns]

### Train - Test Split

In [31]:
X = data.loc[:, data.columns != 'price_doc']
y = data[['price_doc']]   #Target Feature

In [32]:
# Splitting Dataset with 70%-30% ratio.

trainX, testX, trainy, testy = train_test_split(X, y, test_size=0.3, random_state=2)
print(trainX.shape)
print(trainy.shape)
print(testX.shape)
print(testy.shape)

(69999, 270)
(69999, 1)
(30000, 270)
(30000, 1)


### Min - Max Scaling

In [33]:
from sklearn.preprocessing import MinMaxScaler
scaling = MinMaxScaler()

scaling.fit(trainX)
trainX_scaled = scaling.transform(trainX)
testX_scaled = scaling.transform(testX)

In [34]:
trainX_scaled = pd.DataFrame(trainX_scaled, columns= trainX.columns)
testX_scaled = pd.DataFrame(testX_scaled, columns= testX.columns)

In [35]:
trainX_scaled.head()

full_sq   life_sq     floor  product_type    area_m  raion_popul  \
0  0.421856  0.865725  0.756344           0.0  0.044077     0.484455   
1  0.006197  0.002675  0.012987           1.0  0.035300     0.576161   
2  0.011456  0.003477  0.207796           0.0  0.051482     0.718295   
3  0.009729  0.003940  0.117864           1.0  0.030936     0.522001   
4  0.007020  0.002458  0.128645           1.0  0.039047     0.300543   

   green_zone_part  indust_part  children_preschool  \
0         0.993961     0.391709            0.592662   
1         0.177899     0.244077            0.429494   
2         0.226573     0.133661            0.677866   
3         0.498819     0.004238            0.357255   
4         0.081285     0.471445            0.191149   

   preschool_education_centers_raion  ...  cafe_count_5000_price_1500  \
0                           0.744457  ...                    0.466566   
1                           0.461538  ...                    0.034321   
2                           0.692308  ...                    0.028081   
3                           0.384615  ...                    0.048362   
4                           0.384615  ...                    0.040785   

   cafe_count_5000_price_2500  cafe_count_5000_price_4000  \
0                    0.789084                    0.347580   
1                    0.026525                    0.006803   
2                    0.029178                    0.013605   
3                    0.035709                    0.000000   
4                    0.029558                    0.020408   

   cafe_count_5000_price_high  big_church_count_5000  church_count_5000  \
0                    0.553980               0.430521           0.152947   
1                    0.000000               0.046358           0.072000   
2                    0.033333               0.033113           0.016000   
3                    0.000000               0.033113           0.108000   
4                    0.000000               0.050456           0.108951   

   mosque_count_5000  leisure_count_5000  sport_count_5000  market_count_5000  
0           0.161695            0.837178          0.696305           0.353020  
1           0.500000            0.009434          0.197248           0.333429  
2           0.000000            0.009434          0.146789           0.238164  
3           0.365578            0.018868          0.177320           0.298602  
4           0.190550            0.073228          0.266279           0.601073  

[5 rows x 270 columns]

## Linear Regression Model

In [36]:
plr = LinearRegression().fit(trainX_scaled,trainy)
Y_test_pred = plr.predict(testX_scaled)

print("R Square Score",plr.score(testX_scaled,testy))

R Square Score 0.6196553606535521


In [37]:
1 - (1-plr.score(trainX, trainy))*(len(trainy)-1)/(len(trainy)-trainX.shape[1]-1)

-7737057216972.518

In [38]:
from sklearn.metrics import mean_squared_error, r2_score
print("Mean squared error: %.2f" % mean_squared_error(testy, Y_test_pred))
print("Coefficient of determination: %.3f" % r2_score(testy, Y_test_pred))

Mean squared error: 182322811098152.19
Coefficient of determination: 0.620


## Feature Selection

In [39]:
from sklearn.feature_selection import SequentialFeatureSelector
regress = LinearRegression()
sfs = SequentialFeatureSelector(regress, direction = 'forward',n_features_to_select = 20 )
sfs.fit(X,y)
print(sfs.get_feature_names_out())

['full_sq' 'floor' 'build_count_monolith' 'industrial_km'
 'public_transport_station_min_walk' 'fitness_km' 'hospice_morgue_km'
 'university_km' 'prom_part_500' 'cafe_count_500_price_1000'
 'cafe_count_500_price_4000' 'mosque_count_500' 'leisure_count_500'
 'green_part_1000' 'office_sqm_1000' 'cafe_count_1000_price_high'
 'leisure_count_1000' 'market_count_1000' 'sport_count_2000'
 'office_sqm_5000']


Saving these 20 features in variables

In [40]:
data_20_forward = data[sfs.get_feature_names_out()]
data_20_forward

full_sq      floor  build_count_monolith  industrial_km  \
0        43.000000   4.000000              2.000000       1.080934   
1        34.000000   3.000000              4.000000       0.966479   
2        43.000000   2.000000              4.000000       0.939275   
3        77.000000   4.000000             16.000000       0.392871   
4        67.000000  14.000000              2.000000       0.256487   
...            ...        ...                   ...            ...   
99994    79.199672   9.843592              2.000000       0.372054   
99995    76.494667  15.494667              2.000000       0.262782   
99996  4449.811193  36.753485              6.886417      13.080155   
99997    39.497706   3.143513              2.000000       0.486071   
99998    39.186180   6.255281              2.000000       0.359358   

       public_transport_station_min_walk  fitness_km  hospice_morgue_km  \
0                               3.299822    0.485841           2.392353   
1                               0.783160    0.668364           2.543747   
2                               3.945073    0.733101           0.621357   
3                               0.857764    0.220288           1.864637   
4                               2.270726    0.810413           3.391117   
...                                  ...         ...                ...   
99994                           2.662923    0.522193           1.017399   
99995                           1.863938    0.621944           0.955784   
99996                          47.796476   10.060075          35.127832   
99997                           0.476080    0.679855           0.588024   
99998                           2.454056    0.532110           0.477162   

       university_km  prom_part_500  cafe_count_500_price_1000  \
0           6.715026       0.000000                   0.000000   
1           1.852560       0.000000                   3.000000   
2           0.841254       0.000000                   2.000000   
3           0.991683       4.440000                  10.000000   
4           3.830021      19.420000                   1.000000   
...              ...            ...                        ...   
99994       1.820031       7.998808                   0.000000   
99995       1.873918      11.168354                   1.996953   
99996      35.381753       3.736613                  35.745895   
99997       1.617066       0.790755                   0.000000   
99998       2.254138       2.384493                   1.627641   

       cafe_count_500_price_4000  mosque_count_500  leisure_count_500  \
0                       0.000000          0.000000           0.000000   
1                       1.000000          0.000000           0.000000   
2                       0.000000          0.000000           0.000000   
3                       0.000000          0.000000           2.000000   
4                       0.000000          0.000000           0.000000   
...                          ...               ...                ...   
99994                   0.000000          0.000000           0.000000   
99995                   0.000000          0.000000           0.000000   
99996                  12.827928          0.534515           3.115388   
99997                   0.000000          0.000000           0.000000   
99998                   0.000000          0.000000           0.000000   

       green_part_1000  office_sqm_1000  cafe_count_1000_price_high  \
0             7.360000     3.050000e+04                    0.000000   
1            26.660000     8.660000e+04                    0.000000   
2             4.990000     0.000000e+00                    0.000000   
3             3.340000     4.209520e+05                    1.000000   
4             0.000000     2.751350e+05                    0.000000   
...                ...              ...                         ...   
99994        16.181234     5.930076e+04                    0.000000   
99995         7.059371     1.70

In [41]:
X = data_20_forward
y = data[['price_doc']]   #Target Feature

In [42]:
#Deviding into Train - Test after selecting features using foward feature selection. 

trainX, testX, trainy, testy = train_test_split(X, y, test_size=0.3, random_state=2)
print(trainX.shape)
print(trainy.shape)
print(testX.shape)
print(testy.shape)

(69999, 20)
(69999, 1)
(30000, 20)
(30000, 1)


## Linear Regression After Feature Selection.

In [44]:
#Again applying Linear Regression after doing foward selection.

plr = LinearRegression().fit(trainX,trainy)
Y_test_pred = plr.predict(testX)

print("R Square Score",plr.score(testX,testy))
print("Adjusted R-Square", 1 - (1-plr.score(trainX, trainy))*(len(trainy)-1)/(len(trainy)-trainX.shape[1]-1))
print("Mean squared error: %.2f" % mean_squared_error(testy, Y_test_pred))
print("Coefficient of determination: %.3f" % r2_score(testy, Y_test_pred))


R Square Score 0.6213770095458472
Adjusted R-Square 0.6194448722252192
Mean squared error: 181497517842260.28
Coefficient of determination: 0.621


As we see, Performance of model increases after doing feature selection.

The Model is giving R-Square value of 0.62 which is quite fine at this size of data, but it can be improve by using other model like random forest regressor or GBR, but linear regression is also good.